File: backend/app/schemas/client_ai_knowledge.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator, model_validator


class ClientAiConversationMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str = Field(min_length=1, max_length=1000)

    @field_validator("content")
    @classmethod
    def normalize_content(cls, value: str) -> str:
        return " ".join(value.split())


class ClientAiAskRequest(BaseModel):
    question: str = Field(min_length=2, max_length=1000)
    conversation: list[ClientAiConversationMessage] = Field(
        default_factory=list,
        max_length=8,
    )

    @field_validator("question")
    @classmethod
    def normalize_question(cls, value: str) -> str:
        return " ".join(value.split())

    @model_validator(mode="after")
    def bound_conversation(self):
        if sum(len(item.content) for item in self.conversation) > 4000:
            raise ValueError("Conversation history is too large")
        return self


class ClientAiSource(BaseModel):
    source_type: Literal[
        "client", "email", "document", "project", "inspection", "timeline"
    ]
    source_id: int
    title: str
    date: datetime | None = None
    route: str
    snippet: str = Field(max_length=600)


class ClientAiCoverage(BaseModel):
    structured_fields: int = 0
    projects_considered: int = 0
    inspections_considered: int = 0
    emails_searched: int = 0
    documents_lexical_searched: int = 0
    document_vectors_used: int = 0
    timeline_events_considered: int = 0


class ClientAiAskResponse(BaseModel):
    answer: str
    sources: list[ClientAiSource]
    coverage: ClientAiCoverage
    semantic_status: Literal["available", "unavailable", "not_used", "limited"]
    limitations: list[str]
    direct_answer: bool = False
    model: str | None = None

File: frontend/lib/features/client_ai_knowledge/domain/client_ai_knowledge.dart
class ClientAiSource {
  const ClientAiSource({
    required this.sourceType,
    required this.sourceId,
    required this.title,
    required this.route,
    required this.snippet,
    this.date,
  });
  factory ClientAiSource.fromJson(Map<String, dynamic> json) => ClientAiSource(
    sourceType: json['source_type'] as String,
    sourceId: json['source_id'] as int,
    title: json['title'] as String,
    route: json['route'] as String,
    snippet: json['snippet'] as String,
    date: json['date'] == null ? null : DateTime.parse(json['date'] as String),
  );
  final String sourceType;
  final int sourceId;
  final String title;
  final String route;
  final String snippet;
  final DateTime? date;
}

class ClientAiCoverage {
  const ClientAiCoverage({
    required this.emailsSearched,
    required this.documentsLexicalSearched,
    required this.documentVectorsUsed,
    required this.projectsConsidered,
    required this.inspectionsConsidered,
    required this.timelineEventsConsidered,
  });
  factory ClientAiCoverage.fromJson(Map<String, dynamic> json) =>
      ClientAiCoverage(
        emailsSearched: json['emails_searched'] as int? ?? 0,
        documentsLexicalSearched:
            json['documents_lexical_searched'] as int? ?? 0,
        documentVectorsUsed: json['document_vectors_used'] as int? ?? 0,
        projectsConsidered: json['projects_considered'] as int? ?? 0,
        inspectionsConsidered: json['inspections_considered'] as int? ?? 0,
        timelineEventsConsidered:
            json['timeline_events_considered'] as int? ?? 0,
      );
  final int emailsSearched;
  final int documentsLexicalSearched;
  final int documentVectorsUsed;
  final int projectsConsidered;
  final int inspectionsConsidered;
  final int timelineEventsConsidered;
}

class ClientAiAnswer {
  const ClientAiAnswer({
    required this.answer,
    required this.sources,
    required this.coverage,
    required this.semanticStatus,
    required this.limitations,
    required this.directAnswer,
    this.model,
  });
  factory ClientAiAnswer.fromJson(Map<String, dynamic> json) => ClientAiAnswer(
    answer: json['answer'] as String,
    sources: (json['sources'] as List<dynamic>? ?? const <dynamic>[])
        .map((item) => ClientAiSource.fromJson(item as Map<String, dynamic>))
        .toList(growable: false),
    coverage: ClientAiCoverage.fromJson(
      json['coverage'] as Map<String, dynamic>? ?? const <String, dynamic>{},
    ),
    semanticStatus: json['semantic_status'] as String,
    limitations: (json['limitations'] as List<dynamic>? ?? const <dynamic>[])
        .cast<String>(),
    directAnswer: json['direct_answer'] as bool? ?? false,
    model: json['model'] as String?,
  );
  final String answer;
  final List<ClientAiSource> sources;
  final ClientAiCoverage coverage;
  final String semanticStatus;
  final List<String> limitations;
  final bool directAnswer;
  final String? model;
}
